### DISCUSSION TITLE
[Proposal] Parameter pattern checking/assertions

### DISCUSSION THREAD

**Discussion by IS4Code** [dotnet/csharplang#6414] [Language Ideas]
Sorry if this was proposed before. This is an extension of null parameter checking, to allow other kinds of checks than just for null, using a syntax similar to #6100 but taking advantage of the existing pattern matching facilities.

### Demonstration

```cs
static void Method(object obj, int i)
    assert obj is not null
    assert i is >= 0 throw "The argument must be positive."
{

}

// translates into

static void Method(object obj, int i)
{
    if(!(obj is not null)) throw new ArgumentNullException(nameof(obj));
    if(!(i is >= 0)) throw new ArgumentOutOfRangeException(nameof(i), (object)i, "The argument must be positive.");
}
```

### Syntax

The syntax allows for a new `assert` clause (other existing keywords are possible, such as `required`, `checked`, or (with inverted meaning) `throw if`) after the parameter list of a method, in the same position where `where` clauses might appear:

```
assert <param-name> is <pattern> [ throw <throw-expr> ]
```

`param-name` must be a valid parameter name, and `pattern` must be able to match the parameter. `throw-expr` must be an expression, with type assignable to either `System.Expression` or `string` (explained later).

The clause may appear only in places where a method body is generated (in concrete methods, constructors, local functions, lambda expressions, and primary constructors of records). In constructors, it may be placed before or after `: this(…)` or `: base(…)` calls, indicating where the validation should occur relative to the constructor call (in case of the implicit `: this()`, before the call).

Both `pattern` and `throw-expr` may use all the parameters, and `this` if not on a constructor or `: this(…)` or `: base(…)` has already been called. `await` is disallowed even when the method is `async`. New variables may be declared in `pattern`, which are visible (and usable per standard pattern rules) in the following `assert` clauses, `: this(…)` or `: base(…)` calls, and in the method body.

### Semantics

Each `assert` clause is converted into the corresponding `if(!(<param-name> is <pattern>))` statement(s), all of which are placed immediately at the beginning of the method body, in the specific order (and with respect to the placement of the `: this(…)` or `: base(…)` call). In `async` methods and iterators, the statement is executed eagerly, before the corresponding state machine is created, and therefore its side effects or exception are observable immediately, without calling `MoveNext` or awaiting the returned task.

If `throw-expr` is present and is of type `System.Exception`, the clause is represented precisely by `if(!(<param-name> is <pattern>)) throw <throw-expr>;`. Otherwise, the pattern is analyzed using the following algorithm:

1. The pattern is converted to CNF (conjunctive normal form), i.e. the form `(A1 or A2 or …) and (B1 or B2 or …) and …`.
2. The CNF is split into disjunctions (by the `and`), in the form `A1 or A2 or …`.
3. The kind of each pattern in a disjunction is used to determine the exception type and its construction (per the table below).
4. If all the patterns in the disjunction agree on the construction of the exception object, it is used for the disjunction, otherwise the *default* one is selected.
5. Each disjunction corresponds to `if(!(<param-name> is (A1 or A2 or …))) throw …;`.

The type of the individual exception and its construction is determined by the kind of the individual patterns. Here, `throw-expr` is the error message, or `(string)null` if not provided. If a general exception type such as `ArgumentException` or `ArgumentOutOfRangeException` is deduced, the absence of `throw-expr` should be a warning (because the user may expect the compiler to generate something more useful, for example in situations where a pattern that can be deduced to a concrete exception is changed in a way that makes that impossible).

Kind of pattern|Exception constructor
-|-
*Default*|`new System.ArgumentException(<throw-expr> ?? new ArgumentException().Message, nameof(<param-name>))`
`not null` (also implied by any null-checking pattern)|`new System.ArgumentNullException(nameof(<param-name>), <throw-expr>)` or `new System.ArgumentNullException(nameof(<param-name>))` if `throw-expr` is not provided
constant or relational pattern|`new System.ArgumentOutOfRangeException(nameof(<param-name>), <param-name>, <throw-expr> ?? new System.ArgumentOutOfRangeException().Message)`
None at the moment, but may be useful in combination with #3179|`new System.ComponentModel.InvalidEnumArgumentException(nameof(<param-name>), (int)<param-name>, <param-name>.GetType())` or `new System.ComponentModel.InvalidEnumArgumentException(<throw-expr>)` if `throw-expr` is provided.

Additional types of exceptions could be added in the future to accommodate the different types of patterns (such as an exception for an unexpected type, invalid property, list members etc.).

### Example

```cs
public Class(int a, TimeSpan? dt, bool b)
    assert a >= 0
    : base()
    assert dt is { TotalSeconds: > 1 }
    assert b is true throw "b must be true"
{

}

// is equivalent to

public Class(int a, TimeSpan? dt, bool b)
{
    if(a < 0) throw new ArgumentOutOfRangeException(nameof(a), a, new ArgumentOutOfRangeException().Message); // seems there is no way to use the "default" message directly together with ActualValue; warn that a message is not provided
    base();
    if(dt == null) throw new ArgumentNullException(nameof(dt)); // implied by the { } pattern
    if(dt.TotalSeconds <= 1) throw new ArgumentException(new ArgumentException().Message, nameof(dt)); // warn that a message is not provided
    if(b != true) throw new ArgumentOutOfRangeException(nameof(b), b, "b must be true");
}
```

### Further ideas

* Should the compiler enforce that only `ArgumentException` could be thrown? In that case, some other exception could be wrapped in the `ArgumentException` as its `innerException`, to preserve the name of the parameter in all situations.
* Is supporting arbitrary conditions useful, or do patterns cover most of the current situations? If not, something like `assert i on Condition(i) throw …` could be useful to specify an arbitrary condition, but still take advantage of the automatic exception construction. Perhaps `assert str on Int32.TryParse(str, out var num)` would be nifty.
* Combining the two previous ideas, `assert i on try Expression(i)` might catch any exception thrown by `Expression` and wrap it as `innerException`. This might be useful for something like `assert str on try Int32.Parse(str)` except we already have `TryParse`.
* Should the compiler attempt to generate a more descriptive exception message from the pattern, or provide the string representation of the pattern in some way? Something like `throw new ArgumentException("Argument must be { TotalSeconds: > 1 }", nameof(dt))` is not perfect, but could be improved.

**Comment by svick** [comment]
Previous discussion: https://github.com/dotnet/csharplang/issues/105
